### PR TITLE
Extended time series regularizer to support multiseries

### LINKED ISSUE #4304: Extend Time Series Regularizer to support multiseries

### PR DESCRIPTION
**Pull request by MichaelFu512**
### Pull Request Description
Extended time series regularizer to support multiseries
Closes #4304

### Example of what this does
Let's say you have an unstacked X:
 
Screen Shot 2023-09-12 at 12 21 07 PM

Where "2018-01-11", "2018-01-12" and "2018-01-16" are missing.

And you have an unstacked y:
Screen Shot 2023-09-12 at 12 23 18 PM

Which is missing the values that correspond to the missing dates from X_unstacked (values jumps from 49 -> 60 and 74 -> 80).

If you call TimeSeriesRegularizer on X and y:
Screen Shot 2023-09-12 at 12 25 37 PM

y will now regularize into:

Screen Shot 2023-09-13 at 12 39 02 PM

Works also for duplicate datetimes, extra datetimes, and misaligned datetimes.

-----
*After creating the pull request: in order to pass the **release_notes_updated** check you will need to update the "Future Release" section of* `docs/source/release_notes.rst` *to include this pull request by adding :pr:`123`.*

### CODE REVIEW

**Comment by machineFL** [comment]
No linked issues found. Please add the corresponding issues in the pull request description. <br/> [Use GitHub automation to close the issue when a PR is merged](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

File: evalml/pipelines/components/transformers/preprocessing/time_series_regularizer.py [hunk 1]
```diff
@@ -295,7 +296,13 @@ def transform(self, X, y=None):
...
         cleaned_y = None
         if y is not None:
-            y_dates = pd.DataFrame({self.time_index: X[self.time_index], "target": y})
+            if isinstance(y, pd.Series):
+                y_dates = pd.DataFrame(
+                    {self.time_index: X[self.time_index], "target": y},
+                )
+            else:
+                y_dates = y
+                y_dates[self.time_index] = X[self.time_index]
```
Comment (MichaelFu512) [inline evalml/pipelines/components/transformers/preprocessing/time_series_regularizer.py:305]: Can y_unstacked have a column for time index or should it not?

File: evalml/tests/component_tests/test_time_series_regularizer.py [hunk 2]
```diff
@@ -325,3 +325,36 @@ def test_ts_regularizer_uneven(
 
     error_dict = ts_regularizer.error_dict
     assert_features_and_length_equal(X, y, X_output, y_output, error_dict)
+
+
+@pytest.mark.parametrize(
+    "missing",
+    [[], [10], [10, 11], [10, 11, 15]],
+)
+def test_ts_regularizer_multiseries(multiseries_ts_data_unstacked, missing):
```
Comment (eccabay) [inline evalml/tests/component_tests/test_time_series_regularizer.py:334]: I think we need more robust testing here - there are more scenarios that will need the regularizer than just missing values. Can you parametrize the existing tests to cover the multiseries case?

**Review by eccabay — CHANGES_REQUESTED** [review]
We're going to need more testing, but the impl itself is solid!

**Review by jeremyliweishih — APPROVED** [review]
Quick refactor but otherwise LGTM.

**Review by christopherbunn — APPROVED** [review]

File: evalml/tests/component_tests/test_time_series_regularizer.py [hunk 3]
```diff
@@ -23,13 +23,29 @@ def get_df(dates):
...
+
+    reg_X["dates"] = dates
+    reg_X["feature_a_0"] = [int(i) for i in range(len(dates))]
+    reg_X["feature_a_1"] = [int(i) for i in range(len(dates), 0, -1)]
+    reg_X["feature_b_0"] = [int(i) for i in range(len(dates) * 2, 0, -2)]
+    reg_X["feature_b_1"] = [int(i) for i in range(0, len(dates) * 2, 2)]
+
+    reg_y = pd.DataFrame()
+    reg_y["target_0"] = [int(i) for i in range(len(dates))]
+    reg_y["target_1"] = [int(i) for i in range(len(dates), 0, -1)]
```
Comment (eccabay) [inline evalml/tests/component_tests/test_time_series_regularizer.py:37]: You can simplify all the `[int(i) for i in range(<params>)]` to be just `list(range(<params>))`!

File: evalml/tests/component_tests/test_time_series_regularizer.py [hunk 4]
```diff
@@ -60,16 +76,47 @@ def assert_features_and_length_equal(
...
         ].iloc[0]
         pd.testing.assert_series_equal(
             pd.Series(input_feat.values),
-            pd.Series(outout_feat.values),
+            pd.Series(output_feat.values),
+            check_dtype=check_dtype,
         )
 
 
+def check(is_multiseries, X, y, X_output, y_output, error_dict):
```
Comment (eccabay) [inline evalml/tests/component_tests/test_time_series_regularizer.py:90]: Can the function name be slightly more descriptive? It's not immediately obvious *what* is being checked here

**Review by eccabay — APPROVED** [review]
A few small comments, but nothing blocking. The tests are much cleaner now, thank you!

**Pull request merged**